DigitalLearningSolutions.Web.Tests/Controllers/ForgotPassword/ForgotPasswordControllerTests.cs

```diff
@@ -0,0 +1,75 @@
+﻿namespace DigitalLearningSolutions.Web.Tests.Controllers.ForgotPassword
+{
+    using DigitalLearningSolutions.Data.Exceptions;
+    using DigitalLearningSolutions.Data.Services;
+    using DigitalLearningSolutions.Web.Controllers;
+    using DigitalLearningSolutions.Web.ViewModels.ForgotPassword;
+    using FakeItEasy;
+    using FluentAssertions;
+    using FluentAssertions.AspNetCore.Mvc;
+    using Microsoft.Extensions.Logging;
+    using NUnit.Framework;
+
+    class ForgotPasswordControllerTests
+    {
+        private ForgotPasswordController controller;
+        private IPasswordResetService passwordResetService;
+
+        [SetUp]
+        public void SetUp()
+        {
+            passwordResetService = A.Fake<IPasswordResetService>();
+
+            controller = new ForgotPasswordController(passwordResetService);
+        }
+
+        [Test]
+        public void No_model_should_render_basic_form()
+        {
+            // When
+            var result = controller.Index(model: null);
+
+            // Then
+            result.Should().BeViewResult().WithDefaultViewName();
+        }
+
+        [Test]
+        public void Successful_email_submission_should_render_confirm_page()
+        {
+            // Given
+            A.CallTo(() => passwordResetService.SendResetPasswordEmail(A<string>._)).DoesNothing();
+
+            // When
+            var result = controller.Index("recipient@example.com");
+
+            // Then
+            result.Should().BeRedirectToActionResult().WithActionName("Confirm");
+        }
+
+        [Test]
+        public void Email_submission_should_call_password_reset_service()
+        {
+            // When
+            controller.Index("recipient@example.com");
+
+            // Then
+            A.CallTo(() => passwordResetService.SendResetPasswordEmail(A<string>._)).MustHaveHappened();
+        }
+
+        [Test]
+        public void Bad_email_submission_should_render_basic_form_with_error()
+        {
+            string errorMessage = "No user account could be found with the specified email address";
+
+            // Given
+            A.CallTo(() => passwordResetService.SendResetPasswordEmail(A<string>._)).Throws(new EmailAddressNotFoundException(errorMessage));
```

Comment:
Should we add an equivalent test for the db error case? Just because in theory we could handle these errors differently?

Reply:
Sorry, which DB error case specifically? Currently, any exception that isn't explicitly handled inside the controller/services causes the user to get redirected to the UnknownError page, the testing for which is done in LearningSolutionsControllerTests. As there isn't much useful information we could supply a user for this error, I can't think of a more specific way to handle it.

Reply:
Oh, I meant the ResetPasswordInsertException, so an equivalent test would just be:
A.CallTo(() => passwordResetService.SendResetPasswordEmail(A<string>._)).Throws(new ResetPasswordInsertException(errorMessage));